lib/temporal/data_converter.rb

```diff
@@ -1,40 +1,73 @@
+require 'temporal/api/common/v1/message_pb'
+require 'temporal/errors'
 require 'temporal/payload_converter'
 
 module Temporal
   class DataConverter
-    def initialize(payload_converter: Temporal::PayloadConverter::DEFAULT)
+    class MissingPayload < Temporal::Error; end
+
+    def initialize(payload_converter: Temporal::PayloadConverter::DEFAULT, payload_codecs: [])
       @payload_converter = payload_converter
+      @payload_codecs = payload_codecs
     end
 
     def to_payloads(data)
       return if data.nil? || Array(data).empty?
 
       payloads = Array(data).map { |value| to_payload(value) }
-      Temporal::Api::Common::V1::Payloads.new(payloads: payloads)
+      Temporal::Api::Common::V1::Payloads.new(payloads: encode(payloads))
     end
 
     def to_payload_map(data)
-      data.to_h { |key, value| [key.to_s, to_payload(value)] }
+      data.to_h do |key, value|
+        payload = to_payload(value)
+        encoded_payload = encode([payload]).first
+        raise MissingPayload, 'Payload Codecs returned no payloads' unless encoded_payload
+
+        [key.to_s, encoded_payload]
+      end
     end
 
     def from_payloads(payloads)
       return unless payloads
 
-      payloads.payloads.map { |payload| from_payload(payload) }
+      decode(payloads.payloads)
+        .map { |payload| from_payload(payload) }
     end
 
     def from_payload_map(payload_map)
       return unless payload_map
 
       # Protobuf's Hash isn't compatible with the native Hash, ignore rubocop here
-      # rubocop:disable Style/HashTransformValues, Style/MapToHash
-      payload_map.map { |key, value| [key, from_payload(value)] }.to_h
-      # rubocop:enable Style/HashTransformValues, Style/MapToHash
+      # rubocop:disable Style/MapToHash
+      payload_map.map do |key, payload|
+        decoded_payload = decode([payload]).first
+        raise MissingPayload, 'Payload Codecs returned no payloads' unless decoded_payload
+
+        [key, from_payload(decoded_payload)]
+      end.to_h
+      # rubocop:enable Style/MapToHash
     end
 
     private
 
-    attr_reader :payload_converter
+    attr_reader :payload_converter, :payload_codecs
```

Comment:
I wonder if you want to make `payload_codec` singular and have a composite if needed.

Also, I wonder if these attrs should be public. Is it reasonable for someone to just want to call the conversion part or the encoding part of their data converter if they want? Or easily obtain the default payload converter of the default data converter without having to know what class that is?

Reply:
Not sure what's the value of a `Composite` codec where one can be avoided — it's just more internals for the users to interact with. Btw, I think a similar approach can be taken with the payload converters, however we've agreed to keep it to preserve consistency with other SDKs.

As for the public access to these attrs — what would be the practical value of it? Keep in mind that the `DataConverter` is not something that will be directly accessible by the users

Reply:
> Not sure what's the value of a Composite codec where one can be avoided — it's just more internals for the users to interact with.

Arguably allowing the user to just pass in a single codec instead of multiple is easier.

> Keep in mind that the DataConverter is not something that will be directly accessible by the users

I didn't know this, so nevermind here. Arguably a user should be allowed to convert values using the high level interface instead of us hiding it only for ourselves. But can expose later if requested.